ccl/changefeedccl: TestChangefeedTopicNames failed

ccl/changefeedccl.TestChangefeedTopicNames [failed](https://mesolite.cluster.engflow.com/invocations/default/287d9343-25eb-44b0-bb30-4ef7cbc8eb33?testReportRun=3&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [299240d06a0bf6036c92ca18da148255342fa516](https://github.com/cockroachdb/cockroach/commits/299240d06a0bf6036c92ca18da148255342fa516):


```
=== RUN   TestChangefeedTopicNames
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestChangefeedTopicNames747227876
    test_log_scope.go:82: use -show-logs to present logs inline
    changefeed_test.go:11353: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedTopicNames747227876
--- FAIL: TestChangefeedTopicNames (45.38s)
=== RUN   TestChangefeedTopicNames/pubsub
    helpers_test.go:1394: making server as system tenant
    helpers_test.go:1470: making pubsub feed factory
    changefeed_test.go:11349: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:279
        	            				pkg/ccl/changefeedccl/helpers_test.go:498
        	            				pkg/ccl/changefeedccl/changefeed_test.go:11349
        	            				pkg/ccl/changefeedccl/helpers_test.go:1656
        	            				pkg/ccl/changefeedccl/helpers_test.go:1692
        	Error:      	Received unexpected error:
        	            	operation "withTimeout-1110475433624076289" timed out after 30s (given timeout 30s): stopping job due to TestFeed timeout
        	            	(1) operation "withTimeout-1110475433624076289" timed out after 30s (given timeout 30s)
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func3
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:3046
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func1
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:3035
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) stopping job due to TestFeed timeout
        	            	Error types: (1) *timeutil.TimeoutError (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestChangefeedTopicNames/pubsub
--- FAIL: TestChangefeedTopicNames/pubsub (45.38s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedTopicNames.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54857

ccl/changefeedccl.TestChangefeedTopicNames [failed](https://mesolite.cluster.engflow.com/invocations/default/4f115889-95df-4772-a11f-2e790238280b?testReportRun=2&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [ac68bcaf4a547b9d24127a6da8ee01df7753fea1](https://github.com/cockroachdb/cockroach/commits/ac68bcaf4a547b9d24127a6da8ee01df7753fea1):


```
=== RUN   TestChangefeedTopicNames
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestChangefeedTopicNames1701825067
    test_log_scope.go:82: use -show-logs to present logs inline
    changefeed_test.go:11353: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedTopicNames1701825067
--- FAIL: TestChangefeedTopicNames (46.06s)
=== RUN   TestChangefeedTopicNames/pubsub
    helpers_test.go:1394: making server as system tenant
    helpers_test.go:1470: making pubsub feed factory
    changefeed_test.go:11349: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:279
        	            				pkg/ccl/changefeedccl/helpers_test.go:498
        	            				pkg/ccl/changefeedccl/changefeed_test.go:11349
        	            				pkg/ccl/changefeedccl/helpers_test.go:1656
        	            				pkg/ccl/changefeedccl/helpers_test.go:1692
        	Error:      	Received unexpected error:
        	            	operation "withTimeout-1110757832692563969" timed out after 30s (given timeout 30s): stopping job due to TestFeed timeout
        	            	(1) operation "withTimeout-1110757832692563969" timed out after 30s (given timeout 30s)
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func3
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:3046
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func1
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:3035
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) stopping job due to TestFeed timeout
        	            	Error types: (1) *timeutil.TimeoutError (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestChangefeedTopicNames/pubsub
--- FAIL: TestChangefeedTopicNames/pubsub (46.06s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #154306 ccl/changefeedccl: TestChangefeedTopicNames failed [A-cdc C-test-failure O-robot T-cdc branch-release-25.4 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedTopicNames.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl.TestChangefeedTopicNames [failed](https://mesolite.cluster.engflow.com/invocations/default/da187a2c-8a9b-4a25-a4a7-223c01b4b112?testReportRun=2&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [2e14011ff0af49f901a8347f85c4246bd179d119](https://github.com/cockroachdb/cockroach/commits/2e14011ff0af49f901a8347f85c4246bd179d119):


```
=== RUN   TestChangefeedTopicNames
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestChangefeedTopicNames3782488865
    test_log_scope.go:82: use -show-logs to present logs inline
    changefeed_test.go:11407: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedTopicNames3782488865
--- FAIL: TestChangefeedTopicNames (46.90s)
=== RUN   TestChangefeedTopicNames/pubsub
    helpers_test.go:1394: making server as system tenant
    helpers_test.go:1470: making pubsub feed factory
    helpers_test.go:1198: did not force enriched envelope for CREATE CHANGEFEED FOR "ta%pble_1", "table_2", "table_3", "taBle_4", "table_5", "table_6", "table_7", "table_8", "table%q_9", "table_10", "ta😡b""le_11", "ta""ble_12", "table_13", "table_14", "table_15", "table""_16", "table_17", "table_18", "table_19", "table_20", "table_21", "ta%pble_22", "table_23", "tab
        l e_24", "table_25", "table_26", "table_27", "table_28", "Table😜_29", "t'a&ble_30", "tab\\U0010C648le_31", "t�able_32", "table_33", "tA.ble_34", "table_35", "tab(le_36", "table_37", "table_38", "tabl e_39", "table_40", "table	_41", "table_42", "'table_43", " table*_44", "table\\x62_45", "table_46", "table_47", "table_48", "table_49", "table_50", "table_51", "table_52", "table_53", "table_54", "table_55", "table_56", """tabl\re_57", "table_58", "t'able_59", "table_60", "table_61", "table'_62", "table_63", "�table_64", "table_65", "tab le_66", "table_67", "t
able""_68", "table_69", "table_70", "table_71", "tab	le_72", "�tAble_73", "tabl😬e _74", "table_75", "table_76", "tablE_77", "table_78", "table_79", "table_80", "ta b͡le_81", "table_82", "
        table_83", "tAble_84", "\rtable_85", "table_86", "table_87", "table_88", "table_89", "tab le_90", "table_91", "t/abl%qe_92", "table_93", "table_94", "table_95", "t able_96", "table_97", "tab""le_98", "t	able_99", "table_100" because *changefeedccl.pubsubFeedFactory is not supported
    changefeed_test.go:11403: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:279
        	            				pkg/ccl/changefeedccl/helpers_test.go:498
        	            				pkg/ccl/changefeedccl/changefeed_test.go:11403
        	            				pkg/ccl/changefeedccl/helpers_test.go:1656
        	            				pkg/ccl/changefeedccl/helpers_test.go:1692
        	Error:      	Received unexpected error:
        	            	operation "withTimeout-1111894318067318785" timed out after 30.004s (given timeout 30s): stopping job due to TestFeed timeout
        	            	(1) operation "withTimeout-1111894318067318785" timed out after 30.004s (given timeout 30s)
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func3
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:3046
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.stopFeedWhenDone.func1
        	            	  | 	pkg/ccl/changefeedccl/testfeed_test.go:3035
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) stopping job due to TestFeed timeout
        	            	Error types: (1) *timeutil.TimeoutError (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestChangefeedTopicNames/pubsub
--- FAIL: TestChangefeedTopicNames/pubsub (46.90s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #154306 ccl/changefeedccl: TestChangefeedTopicNames failed [A-cdc C-test-failure O-robot P-2 T-cdc branch-release-25.4]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedTopicNames.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/changefeedccl/resolvedspan/frontier.go
+++ b/pkg/ccl/changefeedccl/resolvedspan/frontier.go
@@ -255,18 +255,12 @@ func (f *resolvedSpanFrontier) ForwardResolvedSpan(
 		return false, err
 	}
 	f.latestTS.Forward(r.Timestamp)
-	if r.BoundaryType != jobspb.ResolvedSpan_NONE {
-		newBoundary := resolvedSpanBoundary{
-			ts:  r.Timestamp,
-			typ: r.BoundaryType,
-		}
-		boundaryForwarded := f.boundary.Forward(newBoundary)
-		if boundaryForwarded && !forwarded {
-			// The frontier is considered forwarded if the boundary type
-			// changes to non-NONE and all the spans are at the boundary
-			// timestamp already.
-			forwarded, _, _ = f.AtBoundary()
-		}
+	boundaryForwarded := f.boundary.Forward(r.Timestamp, r.BoundaryType)
+	if boundaryForwarded && !forwarded {
+		// The frontier is considered forwarded if the boundary type
+		// changes to non-NONE and all the spans are at the boundary
+		// timestamp already.
+		forwarded, _, _ = f.AtBoundary()
 	}
 	return forwarded, nil
 }
@@ -408,7 +402,7 @@ func newResolvedSpanBoundary(
 // At returns whether a timestamp is equal to the boundary timestamp
 // and if so, the boundary type as well.
 func (b *resolvedSpanBoundary) At(ts hlc.Timestamp) (bool, jobspb.ResolvedSpan_BoundaryType) {
-	if ts.Equal(b.ts) {
+	if b.IsSet() && ts.Equal(b.ts) {
 		return true, b.typ
 	}
 	return false, 0
@@ -416,17 +410,25 @@ func (b *resolvedSpanBoundary) At(ts hlc.Timestamp) (bool, jobspb.ResolvedSpan_B
 
 // After returns whether the boundary is after a given timestamp.
 func (b *resolvedSpanBoundary) After(ts hlc.Timestamp) bool {
-	return b.ts.After(ts)
+	return b.IsSet() && b.ts.After(ts)
 }
 
 // Forward forwards the boundary to the new boundary if it is later.
 // It returns true if the boundary changed and false otherwise.
-func (b *resolvedSpanBoundary) Forward(newBoundary resolvedSpanBoundary) bool {
-	if newBoundary.After(b.ts) {
-		*b = newBoundary
+func (b *resolvedSpanBoundary) Forward(
+	ts hlc.Timestamp, typ jobspb.ResolvedSpan_BoundaryType,
+) bool {
+	if typ != jobspb.ResolvedSpan_NONE && ts.After(b.ts) {
+		b.ts = ts
+		b.typ = typ
 		return true
 	}
 	return false
+}
+
+// IsSet returns whether the boundary is set.
+func (b *resolvedSpanBoundary) IsSet() bool {
+	return b.typ != jobspb.ResolvedSpan_NONE
 }
 
 // SafeFormat implements the redact.SafeFormatter interface.

--- a/pkg/ccl/changefeedccl/resolvedspan/frontier_test.go
+++ b/pkg/ccl/changefeedccl/resolvedspan/frontier_test.go
@@ -179,13 +179,46 @@ func TestCoordinatorFrontier(t *testing.T) {
 }
 
 type frontier interface {
-	AddSpansAt(startAt hlc.Timestamp, spans ...roachpb.Span) error
+	AddSpansAt(hlc.Timestamp, ...roachpb.Span) error
 	Frontier() hlc.Timestamp
+	Forward(roachpb.Span, hlc.Timestamp) (bool, error)
 	ForwardResolvedSpan(jobspb.ResolvedSpan) (bool, error)
 	InBackfill(jobspb.ResolvedSpan) bool
 	AtBoundary() (bool, jobspb.ResolvedSpan_BoundaryType, hlc.Timestamp)
 	All() iter.Seq[jobspb.ResolvedSpan]
 	Frontiers() iter.Seq2[descpb.ID, span.ReadOnlyFrontier]
+}
+
+func newFrontier(
+	t testing.TB,
+	frontierType string,
+	statementTime hlc.Timestamp,
+	initialHighwater hlc.Timestamp,
+	codec resolvedspan.TableCodec,
+	perTableTracking bool,
+	spans ...roachpb.Span,
+) (frontier, error) {
+	switch frontierType {
+	case "aggregator":
+		return resolvedspan.NewAggregatorFrontier(
+			statementTime,
+			initialHighwater,
+			codec,
+			perTableTracking,
+			spans...,
+		)
+	case "coordinator":
+		return resolvedspan.NewCoordinatorFrontier(
+			statementTime,
+			initialHighwater,
+			codec,
+			perTableTracking,
+			spans...,
+		)
+	default:
+		t.Fatalf("unknown frontier type: %s", frontierType)
+	}
+	panic("unreachable")
 }
 
 func testBackfillSpan(
@@ -379,29 +412,15 @@ func TestFrontierPerTableResolvedTimestamps(t *testing.T) {
 			initialHighWater := hlc.Timestamp{}
 
 			// Create frontier with multiple table spans.
-			f, err := func() (frontier, error) {
-				switch frontierType {
-				case "aggregator":
-					return resolvedspan.NewAggregatorFrontier(
-						statementTime,
-						initialHighWater,
-						codec,
-						true, /* perTableTracking */
-						tableSpans...,
-					)
-				case "coordinator":
-					return resolvedspan.NewCoordinatorFrontier(
-						statementTime,
-						initialHighWater,
-						codec,
-						true, /* perTableTracking */
-						tableSpans...,
-					)
-				default:
-					t.Fatalf("unknown frontier type: %s", frontierType)
-				}
-				panic("unreachable")
-			}()
+			f, err := newFrontier(
+				t,
+				frontierType,
+				statementTime,
+				initialHighWater,
+				codec,
+				true, /* perTableTracking */
+				tableSpans...,
+			)
 			require.NoError(t, err)
 			require.Equal(t, initialHighWater, f.Frontier())
 
@@ -503,29 +522,15 @@ func TestFrontierForwardFullTableSpan(t *testing.T) {
 			statementTime := makeTS(5)
 			var initialHighWater hlc.Timestamp
 
-			f, err := func() (span.Frontier, error) {
-				switch frontierType {
-				case "aggregator":
-					return resolvedspan.NewAggregatorFrontier(
-						statementTime,
-						initialHighWater,
-						codec,
-						true, /* perTableTracking */
-						tableSpans...,
-					)
-				case "coordinator":
-					return resolvedspan.NewCoordinatorFrontier(
-						statementTime,
-						initialHighWater,
-						codec,
-						true, /* perTableTracking */
-						tableSpans...,
-					)
-				default:
-					t.Fatalf("unknown frontier type: %s", frontierType)
-				}
-				panic("unreachable")
-			}()
+			f, err := newFrontier(
+				t,
+				frontierType,
+				statementTime,
+				initialHighWater,
+				codec,
+				true, /* perTableTracking */
+				tableSpans...,
+			)
 			require.NoError(t, err)
 			require.Equal(t, initialHighWater, f.Frontier())
 
@@ -610,26 +615,14 @@ FROM [SHOW RANGES FROM TABLE foo WITH KEYS]`)
 						now := makeTS(timeutil.Now().Unix())
 
 						// Create the frontier and add all the spans.
-						f, err := func() (frontier, error) {
-							switch frontierType {
-							case "aggregator":
-								return resolvedspan.NewAggregatorFrontier(
-									now,
-									now,
-									codec,
-									perTableTracking,
-								)
-							case "coordinator":
-								return resolvedspan.NewCoordinatorFrontier(
-									now,
-									now,
-									codec,
-									perTableTracking,
-								)
-							default:
-								panic("unreachable")
-							}
-						}()
+						f, err := newFrontier(
+							b,
+							frontierType,
+							now,
+							now,
+							codec,
+							perTableTracking,
+						)
 						require.NoError(b, err)
 						require.NoError(b, f.AddSpansAt(now, spans...))
 
@@ -654,4 +647,64 @@ FROM [SHOW RANGES FROM TABLE foo WITH KEYS]`)
 			}
 		}
 	}
+}
+
+func TestFrontierAtBoundary(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	rnd, _ := randutil.NewTestRand()
+	perTableTracking := rnd.Intn(2) == 0
+	t.Logf("per-table tracking: %t", perTableTracking)
+
+	testutils.RunValues(t, "frontier type", []string{"aggregator", "coordinator"},
+		func(t *testing.T, frontierType string) {
+			statementTime := makeTS(timeutil.Now().Unix())
+			var initialHighWater hlc.Timestamp
+			f, err := newFrontier(
+				t,
+				frontierType,
+				statementTime,
+				initialHighWater,
+				mockCodec{},
+				perTableTracking,
+				makeSpan("a", "f"),
+			)
+			require.NoError(t, err)
+
+			// We can't be at boundary until a boundary is set.
+			atBoundary, _, _ := f.AtBoundary()
+			require.False(t, atBoundary)
+			_, err = f.ForwardResolvedSpan(jobspb.ResolvedSpan{
+				Span:      makeSpan("a", "f"),
+				Timestamp: statementTime,
+			})
+			require.NoError(t, err)
+			atBoundary, _, _ = f.AtBoundary()
+			require.False(t, atBoundary)
+
+			// Set a boundary by forwarding part of the span space.
+			ts := statementTime.AddDuration(3 * time.Second)
+			_, err = f.ForwardResolvedSpan(jobspb.ResolvedSpan{
+				Span:         makeSpan("a", "c"),
+				Timestamp:    ts,
+				BoundaryType: jobspb.ResolvedSpan_BACKFILL,
+			})
+			require.NoError(t, err)
+			atBoundary, _, _ = f.AtBoundary()
+			require.False(t, atBoundary)
+
+			// Verify the boundary is reached after forwarding
+			// the rest of the span space.
+			_, err = f.ForwardResolvedSpan(jobspb.ResolvedSpan{
+				Span:         makeSpan("c", "f"),
+				Timestamp:    ts,
+				BoundaryType: jobspb.ResolvedSpan_BACKFILL,
+			})
+			require.NoError(t, err)
+			atBoundary, boundaryType, boundaryTS := f.AtBoundary()
+			require.True(t, atBoundary)
+			require.Equal(t, jobspb.ResolvedSpan_BACKFILL, boundaryType)
+			require.Equal(t, ts, boundaryTS)
+		})
 }

